[Woo POS][Local Catalog] Loading screen for missing catalog when launching POS

staskus opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

WOOMOB-1101

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

1. To show a loading screen for the missing catalog, I implemented a mechanism to track the progress of ongoing sync within `POSCatalogSyncCoordinator`:
- `fullSyncStateStream` to observe ongoing sync state: notStarted, started, completed, error...
- `lastFullSyncState` to check a cached last state if it was completed before

2. Created `POSCatalogLoadingView` to be displayed when the catalog loads. Added matched geometry between a loading view and a catalog view, since it's impossible not to show a normal loading view, at least for a split second. And the progress view position doesn't fall between the two views since the progress view is centered together with the text in the catalog view.

3. Expanded `PointOfSaleDashboardViewHelper` and `PointOfSaleDashboardView.ViewState` to show `catalogSync` state

4. Removed an unnecessary initial delay for syncing the inventory.

Error handling will be done separately:  WOOMOB-1565 

It also made me think of creating a task for caching the eligibility check, and if the POS becomes ineligible, just showing a full-screen overlay. It's an unlikely scenario, and it's better to show POS immediately to give a fast execution impression.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

### Fresh login
1. Fresh login to the Woo app
2. Open the POS immediately after the tab shows up
3. Confirm syncing state shows up
4. Confirm the product list loads once the syncing is done
6. Close POS/the app
7. Reopen and select the POS tab
8. Confirm syncing is no longer shown

### Switch sites
1. Switch to another POS eligible site
2. Open the POS immediately after the tab shows up
3. Confirm syncing state shows up
4. Tap "Exit POS"
5. Wait for sync to complete (observe logs)
6. Open POS
7. Confirm the syncing state is no longer shown, and POS opens

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

Tested on iPad Air 26 simulator.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/user-attachments/assets/b630a4c0-272c-4545-9db4-5ddc038e5f2d

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

staskus committed:
Create posBodySmallBold with underline syle

staskus committed:
Update existing posBodySmallBold usage

staskus committed:
Create POSCatalogLoadingView to show the syncing catalog state

staskus committed:
Allow observing full sync state update stream for POSCatalogSyncCoordinator

Introduces a fullSyncStateStream and state cache to POSCatalogSyncCoordinator, allowing consumers to observe full sync state changes. Adds a POSCatalogSyncState enum to represent sync progress, completion, failure, and never-done states. Refactors sync logic to emit state updates and provides a method to query the last known sync state for a site.

staskus committed:
Update POSCatalogSyncCoordinatorTests

staskus committed:
Remove initial delay for POSCatalogSync when the app launches

staskus committed:
Coordinate animation between loading and catalog loading transition

Generic loading view appears briefly why POS model is loading. It's hard to get away from that. Synchronizing animations between both loading views allows to transition between both loading states without any noticeable transition issues.

staskus committed:
Publish catalogSyncState from PointOfSaleAggregateModel

staskus committed:
Check syncState in PointOfSaleDashboardViewHelper

staskus committed:
Integrate POSCatalogLoadingView in PointOfSaleDashboardView

staskus committed:
Update PointOfSaleDashboardViewHelperTests

staskus committed:
Add a TODO for an error state handling

joshheald reviewed (commented):
> It also made me think of creating a task for caching the eligibility check, and if the POS becomes ineligible, just showing a full-screen overlay. It's an unlikely scenario, and it's better to show POS immediately to give a fast execution impression.

Eligibility for POS, or for local catalog? For the latter, see my latest PR, it moves towards that and has an element of caching. We fall back to the network based approach if they're ineligible. #16276

I've had some initial issues when testing this:

- Catalogs over 1000 items attempt to sync, and never leave the syncing state when it fails. For these sites, we shouldn't even attempt the sync, but this seems to be solved if you merge the PR I mentioned above.
- Syncing catalog can complete, but not transition to the item list, it just spins forever. This happened when I tried to sync an eligible store after opening POS for a store with too many products for local catalog, so I think there may be some state leaked?

joshheald commented on Modules/Sources/PointOfSale/Presentation/CardReaderConnection/UI States/PointOfSaleLoadingView.swift:
```diff
@@ -1,13 +1,25 @@
                 Spacer()
                 ProgressView()
                     .progressViewStyle(POSProgressViewStyle())
+                    .matchedGeometryEffect(id: animation.progressTransitionId, in: animation.namespace, properties: .position)
```
What about using this existing LoadingView for local catalog syncs as well, rather than making a new view? That way we could set text fields to show, but keep using the same loading indicator and avoid visual glitches...

I think it would be a case of adding associated values to the containerState's `loading` enum case, perhaps?

staskus replied:
I think that wouldn;t help with visual glitches.

We transition from PointOfSaleLoadingView within EntryPoint to PointOfSaleLoadingView within Dashboard, and when we add the text, the progress view moves up since, according to the designs, the whole container is centered, not the progress view. I tried centering only the progress view, but that doesn't look right.

But yes, we could keep one view that does more. We would still need to pass matched geometry from the EntryPoint to the Dashboard.

joshheald commented on Modules/Sources/PointOfSale/Models/PointOfSaleAggregateModel.swift:
```diff
@@ -620,6 +623,24 @@ private extension PointOfSaleAggregateModel {
+            for await state in coordinator.fullSyncStateStream {
+                catalogSyncState = state
+            }
+        }
```
I think this task might never end, and therefore be a memory leak. It doesn't capture self, so no retain cycle, but we'll have the task hanging out for each time we leave and enter POS. Perhaps we should keep hold of it, and cancel it when we deinit?

staskus replied:
Yes, it doesn't end since we want an endless stream. Good point on a memory leak. Need to be careful with this 🤔

staskus replied:
`AsyncStream` may not be what I want as a pattern or at least not the way I implemented it. Canceling a task terminates the stream automatically, which has to be recreated. Otherwise, it doesn't emit values anymore. I need a pattern that `Observation` or `Combine` implements for continuous value streaming to multiple listeners.

joshheald replied:
I think it would be fine to just keep hold of the task in a property, and cancel/finish when we deinit the model, wouldn't it?

staskus replied:
Yes, canceling the Task terminates the stream. I think I found a better way just to use `Observation`. I will push the changes soon.

joshheald replied:
Why does that matter, on deinit? We want it to end then, surely.

staskus replied:
We won't use `AsyncStream` anymore but that mattered because when `AsyncStream` is terminated, there was no mechanism to recreate it. It wouldn't emit any new results after opening POS for the second time.

joshheald commented on Modules/Sources/Yosemite/Tools/POS/POSCatalogSyncCoordinator.swift:
```diff
@@ -277,8 +293,65 @@ public actor POSCatalogSyncCoordinator: POSCatalogSyncCoordinatorProtocol {
+        case .syncFailed(let id, _):
+            id
+        case .syncNeverDone(let id):
+            id
```
```suggestion
        case .syncStarted(let id, _), .syncCompleted(let id), .syncFailed(let id, _), .syncNeverDone(let id):
            id
```

joshheald commented on Modules/Sources/Yosemite/Tools/POS/POSCatalogSyncCoordinator.swift:
```diff
@@ -277,8 +293,65 @@ public actor POSCatalogSyncCoordinator: POSCatalogSyncCoordinatorProtocol {
+            return lhsSiteID == rhsSiteID && lhsInitial == rhsInitial
+        case (.syncCompleted(let lhsSiteID), .syncCompleted(let rhsSiteID)):
+            return lhsSiteID == rhsSiteID
+        case (.syncFailed(let lhsSiteID, _), .syncFailed(let rhsSiteID, _)):
```
Are we OK with two failures that have different errors being considered equal?

staskus replied:
Good question. We shouldn't have to error transition, since we would have a loading state in between, but I guess it would be more correct to check error equality as well 🤔

joshheald commented on Modules/Sources/PointOfSale/ViewHelpers/PointOfSaleDashboardViewHelper.swift:
```diff
@@ -1,12 +1,34 @@
+                // TODO: WOOMOB-1565
+                return .error(PointOfSaleErrorState.errorOnLoadingOrders())
+            }
+        }
```
Perhaps we'd avoid the need for this (and some of the async stream stuff) if the GRDBItemsController handled this and set the container state to `.loading(.catalogSyncing)`? I know it's a big change, and I could be missing some fundamental reason why it won't work or would be a bad way to go...

staskus replied:
Fair point. I mean, we would still need changes within the coordinator, since it's the one that does the syncing. Syncing can start at any time, so we need a mechanism that keeps and streams the syncing state.

In this case, I imagine the difference would be replacing aggregateModel+viewHelper with grdbItemsController, but other things would be mostly the same. Correct me if I'm wrong.

> and some of the async stream stuff

Do you see how it can be avoided if we implemented the logic within the controller?

joshheald replied:
> Do you see how it can be avoided if we implemented the logic within the controller?

No, thinking about it more we'd still need something like the stream, which should definitely still be in the coordinator, so that we can show the current sync if there's one already ongoing.

We'd be reading that stream in the GRDBItemsController, and setting the container state there too. I didn't really know that we used this helper, to be honest... it's probably fine in either place.

joshheald reviewed (approved):
This works well, at least when merged with #16276, but I've left some questions/thoughts in line about another way we could approach it which might be easier

staskus commented:
@joshheald Thanks for testing! Yes, I admit I found these tasks with combining sync states, GRDB observation, and SwiftUI more complicated and time-consuming than I anticipated. There are a lot of little details and even after thorough work and testing there are issues that I don't like to see...

> I've had some initial issues when testing this:

> Eligibility for POS, or for local catalog?

Eligibility for POS. I would imagine that with a loaded POS Catalog, we could launch straight into the POS without any loading screen. But that's a topic for future improvements.

> Catalogs over 1000 items attempt to sync, and never leave the syncing state when it fails. For these sites, we shouldn't even attempt the sync, but this seems to be solved if you merge the PR I mentioned above.

Yes, I haven't addressed these limitations in this PR.

> Syncing catalog can complete, but not transition to the item list, it just spins forever. This happened when I tried to sync an eligible store after opening POS for a store with too many products for local catalog, so I think there may be some state leaked?

When developing, I tested multiple times with various stores, but I haven't noticed this case. I will try to understand what could be happening here.

staskus committed:
Update POSCatalogSyncCoordinator.swift

staskus committed:
Check error in syncFailed

staskus committed:
Observe catalog sync state using Observation model

staskus committed:
Move catalogSyncState checking logic to ObservableItemsController

staskus committed:
Merge PointOfSaleLoading view and POSCatalogLoadingView

staskus commented:
@joshheald thanks for the comments. I think I managed to address them with quite substantial changes:

- Moved from using `AsyncStream` to `@Observable` for observing catalog sync state, which removed a category of issues and complications
- Using `coordinator.fullSyncState` in `PointOfSaleObservableItemsController`, which automatically updates the container state if `fullSyncState` changes, not requiring any other observation logic
- Merged `POSCatalogLoadingView` and `PointOfSaleLoadingView` , removed `matchedGeometry` and passing namespace

https://github.com/user-attachments/assets/31c731ba-cd32-4613-9ce6-768bbcdadb34

Assigning to you, to merge after [woomob-1098-woo-poslocal-catalog-add-incremental-sync-triggers-pull-to](https://github.com/woocommerce/woocommerce-ios/tree/woomob-1098-woo-poslocal-catalog-add-incremental-sync-triggers-pull-to)

Thank you again for all the help! 🙇

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/258">23.6</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

joshheald committed:
Merge branch 'trunk' into woomob-1101-woo-poslocal-catalog-loading-screen-for-missing-catalog-when

joshheald committed:
Prevent saving first sync date if it doesn’t happen

joshheald committed:
Start syncs after updating catalog eligibility
